wagtail/images/tests/test_models.py

```diff
@@ -102,6 +117,18 @@ def test_is_stored_locally(self):
     def test_is_stored_locally_with_external_storage(self):
         self.assertFalse(self.image.is_stored_locally())
 
+    @override_settings(
+        DEFAULT_FILE_STORAGE="wagtail.test.dummy_external_storage.DummyExternalStorage"
+    )
+    def test_reopen_based_on_storage_that_is_dynamically_set(self):
+        self.image.file.close()
+        # Simulate the case is witch the storage is set dynamically
+        with mock.patch.object(
+            self.image.file, "storage", AnotherDummyExternalStorage()
+        ):
+            with self.image.open_file() as f:
+                self.assertIs(f.__class__, AnotherDummyExternalStorageFile)
```

Comment:
Nit: Use `assertIsInstance` rather than comparing the class directly
```suggestion
                self.assertIsInstance(f, AnotherDummyExternalStorageFile)
```